File: examples/RUNES_demo/node/expcode.cpp
double expcode(int seg, void* data) {

  Exp_Data *d = (Exp_Data*) data;

  double mintime = INF; // defined in ttkernel.h

  double now = ttCurrentTime();

  

  int k;

  

  for(k=0; k<NBR_AODV; k++) {

    // Invalidate timed out entry(ies) and compute new timer

    if (d->routing_table[k].valid) {

      if ((d->routing_table[k].exptime - now) < EPS) {

	d->routing_table[k].valid = 0;

      } else {

          mintime = min(mintime, d->routing_table[k].exptime);

      }

    }

  }



  if (VERBOSE) {

    mexPrintf("Time: %f timer expiry in Node#%d\n", now, d->nodeID);

  }

    

  // Create new timer if mintime != INF

  if (mintime < INF-100.0) {

    ttCreateTimer("exptimer", mintime, "timer_handler");

    d->expTimer = 1;

    

    if (VERBOSE) {

      mexPrintf("Time: %f Node#%d updating expiry timer to time: %f\n", now, d->nodeID, mintime);

    }

  } else {

    d->expTimer = 0;

  }

     

  return FINISHED;  

}

File: examples/AODV/expcode.m
function [exectime, data] = expcode(seg, data)

global routing_table
global verbose

switch seg,
  
 case 1,
  table = routing_table{data.nodeID};
  mintime = inf;
  noww = ttCurrentTime;

  for k = 1:length(table)
    % Invalidate timed out entry
    if (abs(table{k}.exptime - noww) < 0.0000001)
      table{k}.state = 'invalid';
    end
    if (table{k}.exptime > noww+0.0000001)
      mintime = min(mintime, table{k}.exptime);
    end
  end

  if verbose
    disp(['Time: ' num2str(noww) ' timer expiry in Node#' num2str(data.nodeID)]);
  end
    
  % Create new timer
  if (mintime < inf)
    ttCreateTimer('exptimer', mintime, 'timer_handler');
    data.expTimer = 1;
    
    if verbose
      disp(['Time: ' num2str(noww) ' Node#' num2str(data.nodeID) ' updating expiry timer to time: ' num2str(mintime)]);
    end
      
  else
    data.expTimer = 0;
  end
    
  % Update routing table
  routing_table{data.nodeID} = table;
  
  exectime = -1;
  
end